### PR TITLE
AP_DDS: Copter takeoff service

### PR DESCRIPTION
**Pull request by snktshrma**
Added a service for copter to accept takeoff commands using native DDS support.

### CODE REVIEW

File: Tools/ros2/ardupilot_msgs/srv/Takeoff.srv [hunk 1]
```diff
@@ -0,0 +1,10 @@
+
+# This service requests the vehicle to takeoff
+
+# alt : Set the takeoff altitude
+
+uint8 alt
```
Comment (Ryanf55) [inline Tools/ros2/ardupilot_msgs/srv/Takeoff.srv:6]: Can you change this to a float, matching what the library supports? As a uint8, you can only takeoff to 255m.

File: Tools/ros2/ardupilot_msgs/srv/Takeoff.srv [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+# This service requests the vehicle to takeoff
+
+# alt : Set the takeoff altitude
```
Comment (Ryanf55) [inline Tools/ros2/ardupilot_msgs/srv/Takeoff.srv:4]: Can you specify the altitude datum? Is this height above terrain, height above home, or MSL?

**Review by Ryanf55 — CHANGES_REQUESTED** [review]
Looks pretty good. 

Can you update `libraries/AP_DDS/README.md` for the new service channel?
Also, if you can, is it possible to add a test for it?

File: Tools/ros2/ardupilot_msgs/srv/Takeoff.srv (hunk 1 above)
Comment (snktshrma) [reply Tools/ros2/ardupilot_msgs/srv/Takeoff.srv:6]: I tried with float but somehow service is not accepting given altitude with float and returning 0.00 as commanded altitude

**Comment by snktshrma** [comment]
> Looks pretty good.
> 
> Can you update `libraries/AP_DDS/README.md` for the new service channel? Also, if you can, is it possible to add a test for it?

Alright! I'll add a test for it

**Comment by snktshrma** [comment]
> Looks pretty good.
> 
> Can you update `libraries/AP_DDS/README.md` for the new service channel? Also, if you can, is it possible to add a test for it?

I have added test for takeoff but it is yet to be checked. I'll update once it's tested

**Comment by Ryanf55** [comment]
Can you fix the build errors?
https://github.com/ArduPilot/ardupilot/actions/runs/8877225483/job/24370361196?pr=26911#step:7:2931

**Comment by snktshrma** [comment]
> Can you fix the build errors? https://github.com/ArduPilot/ardupilot/actions/runs/8877225483/job/24370361196?pr=26911#step:7:2931

This build error is fixed @Ryanf55

File: Tools/ros2/ardupilot_msgs/srv/Takeoff.srv (hunk 1 above)
Comment (Ryanf55) [reply Tools/ros2/ardupilot_msgs/srv/Takeoff.srv:6]: Ok. This is a major issue with the PR if we can only command integer resolution altitude and only 255m.

Is the DDS library not accepting the float? Some more information on the failure would be helpful to debug it.

Comment (snktshrma) [reply Tools/ros2/ardupilot_msgs/srv/Takeoff.srv:6]: Yes it seems so. With float, the value that I am passing though ros service command is interpreted as 0.00. So `takeoff_request.alt` is 0.00 in any case.

Comment (snktshrma) [reply Tools/ros2/ardupilot_msgs/srv/Takeoff.srv:6]: @Ryanf55 Any update regarding this float issue?

**Comment by srmainwaring** [comment]
@snktshrma, @Ryanf55 - while working on support for AP in Aerostack2 I examined the PX4 msgs and services for comparison (PX4 added a service interface at the end of last year for vehicle commands). The service request message structure is similar to the mavlink equivalent: https://github.com/PX4/px4_msgs/blob/main/msg/VehicleCommand.msg. Wondering whether we want to use something similar rather than add many separate messages for the various commands?

**Comment by snktshrma** [comment]
hi @srmainwaring !
That seems a feasible idea. It'll be easier for users also to have command of all the required functions.

File: ArduCopter/Copter.h [hunk 3]
```diff
@@ -666,13 +666,13 @@ class Copter : public AP_Vehicle {
                              uint32_t &log_bit) override;
 #if AP_SCRIPTING_ENABLED || AP_EXTERNAL_CONTROL_ENABLED
 #if MODE_GUIDED_ENABLED
+    bool start_takeoff(float alt) override;
```
Comment (Ryanf55) [inline ArduCopter/Copter.h:669]: `bool start_takeoff(const float alt) override;`

^ Can add const

File: libraries/AP_DDS/README.md [hunk 4]
```diff
@@ -253,6 +255,16 @@ requester: making request: ardupilot_msgs.srv.ModeSwitch_Request(mode=4)
 response:
 ardupilot_msgs.srv.ModeSwitch_Response(status=True, curr_mode=4)
 ```
+
+Call the takeoff service:
+
+```bash
+$ ros2 service call /ap/takeoff ardupilot_msgs/srv/takeoff "{alt: 10}"
```
Comment (Ryanf55) [inline libraries/AP_DDS/README.md:262]: It's `ardupilot_msgs/srv/Takeoff` with a capital `T`, otherwise you get:
```
$ ros2 service call /ap/takeoff ardupilot_msgs/srv/takeoff "{alt: 10}"
The passed service type is invalid
```

File: Tools/ros2/ardupilot_msgs/srv/Takeoff.srv (hunk 1 above)
Comment (Ryanf55) [reply Tools/ros2/ardupilot_msgs/srv/Takeoff.srv:6]: I had no problems changing it to a float. Make sure you update both the `srv` and the `.idl` to match. 
image

Comment (Ryanf55) [reply Tools/ros2/ardupilot_msgs/srv/Takeoff.srv:6]: I made a PR to your repo. If you approve, please accept the changes and then fix the history.
https://github.com/snktshrma/ardupilot/pull/1

File: Tools/ros2/ardupilot_dds_tests/ardupilot_dds_tests/copter_takeoff.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+from geopy import point
+from ardupilot_msgs.srv import ArmMotors
+from ardupilot_msgs.srv import ModeSwitch
+from ardupilot_msgs.srv import Takeoff
+
+
+
+COPTER_MODE_GUIDED = 4
+
+TAKEOFF_ALT = 10
```
Comment (Ryanf55) [inline Tools/ros2/ardupilot_dds_tests/ardupilot_dds_tests/copter_takeoff.py:42]:
```suggestion
TAKEOFF_ALT = 10.5
```

**Review by Ryanf55 — CHANGES_REQUESTED** [review]

File: Tools/ros2/ardupilot_msgs/srv/Takeoff.srv (hunk 1 above)
Comment (snktshrma) [reply Tools/ros2/ardupilot_msgs/srv/Takeoff.srv]: I have merged the PR @Ryanf55. I got it working by cleaning the build and rebuilding. Thanks Ryan

**Review by Ryanf55 — APPROVED** [review]
Looks good, tested in SITL and works with floats now.

**Comment by Ryanf55** [comment]
Hi Randy, can you check the copter changes here are good? I tested in SITL, all looks good.

**Comment by Ryanf55** [comment]
Hey there, sorry this didn't get in earlier. Can you fix conflicts? I'm happy with the PR.

**Comment by snktshrma** [comment]
Hi @Ryanf55 ! I have fixed the conflicts.

**Review by Ryanf55 — APPROVED** [review]

**Comment by Ryanf55** [comment]
Awesome, thanks for the quick response. If you can attend either of the dev calls, please request a review of the changes.
Make sure to fix your commits and split them by subsytem.
image


IE, you should have commits for
*  `ArduCopter: `
* `Tools: ros2: 
* `AP_DDS: `
* `AP_Vehicle`

**Comment by snktshrma** [comment]
Got it @Ryanf55 ! I'll try to attend the dev call.
Rest, I'll split the commits as you mentioned.
Thanks

**Comment by Ryanf55** [comment]
The size tests are failing. It might have a conflict, and need a rebase. 
https://github.com/ArduPilot/ardupilot/actions/runs/11228146544/job/31211539090?pr=26911#step:9:108

**Comment by Ryanf55** [comment]
Looks good. The reason the diff looks bad is because you reordered two methods. This introduces noise in the diff.
image

I'm going to force push to put these back to original, which will make it quicker to review/merge. Hopefully during the conference this week.

**Review by Ryanf55 — APPROVED** [review]

**Comment by Ryanf55** [comment]
@snktshrma Now that all the services are wrapped in ifdefs, can you rebase and add that for this PR? 
`AP_DDS_COPTER_TAKEOFF_SERVICE` or something similar could be a good name for it.

**Comment by Ryanf55** [comment]
Hi @snktshrma  Please let me know if you can do this or need someone else. We really want this capability in 4.6.

**Comment by snktshrma** [comment]
> Hi @snktshrma Please let me know if you can do this or need someone else. We really want this capability in 4.6.

Hi @Ryanf55 ! I’m didn’t see your previous comment. I’ll make the change and push the updates asap.

File: libraries/AP_DDS/AP_DDS_Client.cpp [hunk 6]
```diff
@@ -970,11 +1002,6 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
 
         ucdr_init_buffer(&reply_ub, reply_buffer, reply_size);
         const bool serialize_success = rcl_interfaces_srv_SetParameters_Response_serialize_topic(&reply_ub, &set_parameter_response);
-        if (serialize_success == false) {
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.cpp:973]: Oops, unrelated change here, can you undo that?

**Review by Ryanf55 — APPROVED** [review]

**Comment by Ryanf55** [comment]
Please fix the astyle cleanliness. `./Tools/scripts/run_astyle.py`

File: libraries/AP_DDS/AP_DDS_config.h [hunk 7]
```diff
@@ -134,6 +134,10 @@
 #define AP_DDS_MODE_SWITCH_SERVER_ENABLED 1
 #endif
 
+#ifndef AP_DDS_COPTER_TAKEOFF_SERVER_ENABLED
+#define AP_DDS_COPTER_TAKEOFF_SERVER_ENABLED 1
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_config.h:138]: Not a blocker, but maybe we should only enable this if we're on copter?

File: ArduCopter/Copter.cpp [hunk 8]
```diff
@@ -285,13 +285,9 @@ bool Copter::set_target_location(const Location& target_loc)
...
     return mode_guided.set_destination(target_loc);
 }
-#endif //MODE_GUIDED_ENABLED
-#endif //AP_SCRIPTING_ENABLED || AP_EXTERNAL_CONTROL_ENABLED
 
-#if AP_SCRIPTING_ENABLED
-#if MODE_GUIDED_ENABLED
 // start takeoff to given altitude (for use by scripting)
-bool Copter::start_takeoff(float alt)
+bool Copter::start_takeoff(const float alt)
```
Comment (rmackay9) [inline ArduCopter/Copter.cpp:290]: let's not add a "const" here

**Review by rmackay9 — APPROVED** [review]
The Copter changes are tiny so I'm happy.  I haven't reviewed the DDS changes but I trust others will ensure that part is correct

File: libraries/AP_DDS/AP_DDS_config.h (hunk 7 above)
Comment (Ryanf55) [reply libraries/AP_DDS/AP_DDS_config.h:138]: The dev team wants us to change the name to AP_DDS_VTOL_TAKEOFF_SERVER_ENABLED because it works on quadplane too. Is that ok? Otherwise looks good.

File: libraries/AP_DDS/AP_DDS_Service_Table.h [hunk 9]
```diff
@@ -51,6 +54,18 @@ constexpr struct AP_DDS_Client::Service_table AP_DDS_Client::services[] = {
         .reply_topic_name = "rr/ap/mode_switchReply",
     },
 #endif // AP_DDS_MODE_SWITCH_SERVER_ENABLED
+#if AP_DDS_COPTER_TAKEOFF_SERVER_ENABLED
+{
+        .req_id = to_underlying(ServiceIndex::TAKEOFF),
+        .rep_id = to_underlying(ServiceIndex::TAKEOFF),
+        .service_rr = Service_rr::Replier,
+        .service_name = "rs/ap/takeoffService",
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Service_Table.h:62]: Because some disagreement from the dev team on whether this should be relative to home or relative to takeoff altitude, and whether we should be using guided topic, can we add the `experimental` prefix to this so we can change the interface?

File: libraries/AP_DDS/AP_DDS_config.h (hunk 7 above)
Comment (snktshrma) [reply libraries/AP_DDS/AP_DDS_config.h:138]: Got it Ryan! I’ll make the change.

File: libraries/AP_DDS/AP_DDS_Service_Table.h (hunk 9 above)
Comment (snktshrma) [reply libraries/AP_DDS/AP_DDS_Service_Table.h:62]: Hi @Ryanf55 ! You mean "rs/ap/experimental/takeoffService", or AP_DDS_EXPERIMENTAL_ENABLED?

Comment (timtuxworth) [reply libraries/AP_DDS/AP_DDS_Service_Table.h:62]: What about relative to terrain?

Comment (Ryanf55) [reply libraries/AP_DDS/AP_DDS_Service_Table.h:62]: > What about relative to terrain?

The IDL file has it relative to terrain IF rangefinder is in use.

**Comment by peterbarker** [comment]
```
```
Board                    AP_Periph  blimp  bootloader  copter  heli  iofirmware  plane  rover  sub
CubeBlack                           0      *           0       0                 0      0      0
CubeOrange-periph-heavy  *                                                                     
CubeOrangePlus                      0      *           0       0                 0      0      0
CubeRedPrimary                      0      *           0       0                 0      0      0
Durandal                            0      *           0       0                 0      0      0
Hitec-Airspeed           *                                                                     
KakuteH7-bdshot                     0      *           0       0                 0      0      0
MatekF405                           16     *           88      88                8      8      8
Pixhawk1-1M-bdshot                  8                  96      88                8      8      8
f103-QiotekPeriph        *                                                                     
f303-Universal           *                                                                     
iomcu                                                                *                         
revo-mini                           8      *           88      96                8      8      8
skyviper-journey                                       88                                      
skyviper-v2450                                         88                                      
```

... this is costing us bytes.... but DDS is disabled by default.  This isn't good....
```

**Comment by rmackay9** [comment]
this is included in ArduPilot-4.6.0-beta3 at @Ryanf55's request
